crypto/evp/pkey_kdf.c

```diff
@@ -160,9 +188,12 @@ static int pkey_kdf_derive_init(EVP_PKEY_CTX *ctx)
 static int pkey_kdf_derive(EVP_PKEY_CTX *ctx, unsigned char *key,
                            size_t *keylen)
 {
-    EVP_KDF_CTX *kctx = ctx->data;
+    EVP_PKEY_KDF_CTX *pkctx = ctx->data;
+    EVP_KDF_CTX *kctx = pkctx->kctx;
     size_t outlen = EVP_KDF_size(kctx);
 
+    if (!EVP_KDF_CTX_set_params(kctx, pkctx->params))
+        return 0;
```

Comment:
What's the idea with setting the parameters during derive and not before?

Reply:
The TLS1-PRF allows the _seed_ to be set multiple times and these are concatenated (it's not the only one, just the only one where this capability is tested).  The param based API at the other end is expecting all of these to be delivered in a single `OSSL_PARAM` array which means the additional _seed_reset_ call can be avoided -- we don't want params to have side effects and we'd like to minimise the number of across the provider boundary calls.

In the old world, setting the _secret_ had the side effect of calling the _seed_reset_ ctrl and subsequent _seed_ ctrl calls built up the seed.  This seemed a little obtuse and smelt a bit odd.


I'm looking at reworking this using the param builder APIs since the ctrl calls are using construct and the str_ctrl are using your convert from text functionality.  The two are not compatible unfortunately (think about freeing the allocated storage).

I'm thinking about introducing a build from text function and/or some OSSL_PARAM_allocate ones.

Reply:
Essentially, the PKEY APIs are a hodgepodge mess but we've got to continue supporting them.  I feel it is better to have clean new APIs and push the pain onto the legacy side.

Reply:
Thanks for the explanation.  It might be good to add a comment on this...

Reply:
I've added a comment.

Reply:
Does it really need a major rework? We have other places where things can be added multiple times (e.g tls iv or aead aad). That doesnt seem like a big issue to me. In normal operation you wouldnt need the reset unless you actually really wanted to restart.

Reply:
The documentation of the params is quite explicit that ordering is not relevant, with this it is.

Reply:
>The param based API at the other end is expecting all of these to be delivered in a single >OSSL_PARAM array.

I dont get it - If this was being done via controls before - then it was all done one setparams call at a time.

Reply:
Exactly, we don't want to support the old one at a time model.
Better to do everything that's changed in one go.
At the least, it will amortise the extra overhead.

Also consider the special control call that didn't take any arguments, didn't return anything and reset the seed to zero.  Do we really want to replicate such an ugly?

Reply:
Depends on the context.. 
If its supposed to stream then it needs to be seperate calls..

Reply:
Are there any relevant ctrl calls that would impact the output of a PRF?  They are seeding, secrets and nonces.

Reply:
Perhaps, I misunderstood the intention of _stream_ here.

Reply:
I dont know how it gets used, so I was just asking the question.. Maybe it is ok.

Reply:
My guess would be that the concatenation is only used by _libssl_.